File: 프로그래머스/1/42889. 실패율/실패율.py
def solution(N, stages):
    temp = []
    answer = []
    user = 0
    fail_count = [0] * (N + 1)
    
    for stage in stages:
        if stage <= N:
            fail_count[stage] += 1
        user += 1
    
                      
    for i in range(1, N+1):

        if user == 0:
            temp.append((i, 0))
        else:
            temp.append((i, fail_count[i]/user))
            
        user -= fail_count[i]
    
    temp.sort(key = lambda x: -x[1])
        
    return [x[0] for x in temp]

File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.*;

public class Solution {
    public int[] solution(int N, int[] stages) {
        int[] stagesCount = new int[N + 2];
        
        for (int stage: stages) {
            if (stage <= N) {
                stagesCount[stage]++;
            }
        }
        
        List<StageFailRate> failRateList = new ArrayList<>();
        int remainCount = stages.length;
        
        for (int i = 1; i <= N; i++) {
            double failRate = 0.0;
            
            if (remainCount != 0) {
                failRate = (double) stagesCount[i] / remainCount;
            }
            remainCount -= stagesCount[i];
            failRateList.add(new StageFailRate(i, failRate));
            
        }
        
        failRateList.sort(null);
        
        int[] answer = new int[N];
        for (int i = 0; i < N; i++) {
            answer[i] = failRateList.get(i).stage;
        }
        
        return answer;
    }
    
    static class StageFailRate implements Comparable<StageFailRate> {
        int stage;
        double failRate;
        
        public StageFailRate(int stage, double failRate) {
            this.stage = stage;
            this.failRate = failRate;
        }
        
        @Override
        public int compareTo(StageFailRate other) {
            if (this.failRate == other.failRate) {
                return Integer.compare(this.stage, other.stage);
            }
            
            return Double.compare(other.failRate, this.failRate);
        }
    }
}